phantom_token.c

```diff
@@ -665,11 +557,50 @@ static ngx_int_t introspection_response_handler(ngx_http_request_t *request, voi
         return introspection_subrequest_status_code;
     }
 #else
-    jwt_start = request->header_end + sizeof("\r\n") - 1; // FIXME: Won't work if JWT is large
+    jwt_start = request->header_end + sizeof("\r\n") - 1;
+    read_response = true;
 #endif
 
-    size_t jwt_len = request->headers_out.content_length_n;
-    size_t bearer_jwt_len = BEARER_SIZE + jwt_len;
+    jwt_len = request->headers_out.content_length_n;
+    bearer_jwt_len = BEARER_SIZE + jwt_len;
+
+    // When caching is not used, the introspection response is read directly and we must handle large JWTs specially
+    if (read_response)
+    {
+        // We use proxy_pass to call the introspection endpoint, so the ngx_http_proxy_module provides the response to the subrequest.
+        // The response is returned as an upstream buffer and it is common to see code like this, to read headers from the buffer.
+        // - https://github.com/nginx/nginx/blob/master/src/http/modules/ngx_http_proxy_module.c#L1905
+        // - rc = ngx_http_parse_header_line(r, &r->upstream->buffer, 1);
+
+        // It is also common that long headers like cookies get truncated, since the default buffer size is only one memory page.
+        // The standard solution to this problem is to configure an increased proxy_buffer_size.
+        // A default or explicit proxy_buffer_size is always used to handle responses, even when proxy buffering is disabled.
+        // - https://www.uptimia.com/questions/fix-upstream-sent-too-big-header-while-reading-response-header-from-upstream-error
+        // - https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffering
+
+        // The buffer can also contain the response body, which the phantom token module needs to read.
+        // Therefore the module also uses the standard proxy_buffer_size solution to receive large JWTs.
+        // We do this by calculating the buffer size as explained in the development guide.
+        // - https://nginx.org/en/docs/dev/development_guide.html
+
+        // Avoid reading past the end of the buffer if the JWT size exceeds the buffer size
+        buffer_size = ngx_buf_size(&request->upstream->buffer);
```

Comment:
This is the only real change, to add some code to prevent a read past the end of the response buffer when a large JWT is returned from the introspection endpoint. I also added comments to explain the logic and for future reference, since it is a little tricky to understand.